uvicorn/main.py

```diff
@@ -438,6 +437,11 @@ def run(app: typing.Union[ASGIApplication, str], **kwargs: typing.Any) -> None:
         )
         sys.exit(1)
 
+    server = Server(config=config)
+    serve(config, server)
+
+
+def serve(config: Config, server: Server) -> None:
```

Comment:
Out of curiosity, why not simplify the signature here to:
```python
def serve(server: Server) -> None:
    config = server.config
    # ... carry on as usual
```

One other question: why is the `isinstance(app, str)` check happening in `run` rather than `serve`? Seems like this check is now missing if `serve` is called independently of `run`. Perhaps the check should simply be moved to the `Config` constructor?

One final observation: it seems like the method-naming conventions here are getting a bit convoluted & confusing, with: `uvicorn.run` CALLS `uvicorn.serve` CALLS `Server.run` CALLS `Server.serve` (where `Server.serve` and `uvicorn.serve` don't share any particular distinguishing commonality!) If the only reason for calling this method `serve` is because the word `run` has already been used... perhaps it might make sense to keep it more consistent with `uvicorn.run` and call it something like `run_server` instead? (Esp. since the method is ultimately calling `Server.run`, _not_ `Server.serve`!)

Reply:
After further research, it looks like #1012 has the right idea, as it resolves all of the above concerns. I wouldn't change anything from that PR... why not just use that one?

Reply:
I just want `Config` to be passed independently as a parameter. I think both PRs are fine. If yours is merged, I will close this one.

Reply:
> One other question: why is the isinstance(app, str) check happening in run rather than serve? Seems like this check is now missing if serve is called independently of run. Perhaps the check should simply be moved to the Config constructor?

It is a good idea to move this check to Config. I will do it as soon as possible.

Reply:
Why do you want `Config` passed independently? Shouldn't you (always) use the same config object that initialized the`Server`? If so, you could simply do `config = server.config`.

Reply:
There is no deeper reason, I just intuitively think this might be better.

Reply:
I agree we should probably go for `run_server(server)`, as it would be awkward and bug-prone to pass a `config` that doesn't match the server's, right?

Reply:
After thinking about it carefully, you are right. This time my intuition went wrong.